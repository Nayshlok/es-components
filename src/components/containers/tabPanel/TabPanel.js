import React from 'react';
import PropTypes from 'prop-types';
import { colors } from '../../theme';
import styled from 'styled-components';

const SelectedButton = styled.button`
  background-color: ${colors.white};
  border: 1px solid ${colors.gray};
  border-bottom: 1px transparent;
  border-radius: 2px 2px 0 0;
  color: ${colors.black};
  display: inline-block;
  font-size: 18px;
  font-family: Arial, sans-serif;
  margin-right: 2px;
  margin-bottom: -1px;
  padding: 10px 15px;
`;

const TabButton = styled.button`
  background-color: ${colors.white};
  border: 0px;
  border-bottom: 1px;
  color: ${colors.linkColor};
  display: inline-block;
  font-size: 18px;
  font-family: Arial, sans-serif;
  padding: 10px 15px;
  &:hover: {
    background-color: ${colors.grayDarker};
  }
`;

const SelectedDropdown = styled.select`
  background-color: ${colors.white};
  border: 1px solid ${colors.gray};
  border-bottom: 1px transparent;
  border-radius: 2px 2px 0 0;
  box-shadow: 0px;
  color: ${colors.black};
  display: inline-block;
  font-family: Arial, sans-serif;
  font-size: 18px;
  margin-right: 2px;
  padding: 10px 15px;
`;

const OtherDropdown = styled.select`
  background-color: ${colors.white};
  border: 0px;
  border-bottom: 1px;
  color: ${colors.linkColor};
  display: inline-block;
  font-size: 18px;
  font-family: Arial, sans-serif;
  padding: 10px 15px;
  &:hover: {
    background-color: ${colors.grayDarker};
  }
`;

const ContentDiv = styled.div`
  border-top: 1px solid ${colors.grayLight};
`;

function Tab({ name, selected, action, ...props }) {
  return selected ? (
    <SelectedButton>{name}</SelectedButton>
  ) : (
    <TabButton onClick={() => action(name)}>{name}</TabButton>
  );
}

Tab.propTypes = {
  name: PropTypes.string.isRequired,
  selected: PropTypes.bool.isRequired,
  action: PropTypes.func
};

function TabList({ options, name, selected, action, ...props }) {
  const update = event => {
    const value = event.target.value;
    action(value);
  };

  const Title = <option disabled>{name}</option>;

  const selectOptions = options.map(opt => {
    const optionKey = opt.optionText.replace(/\s/g, '');
    return (
      <option key={optionKey} value={optionKey}>
        {opt.optionText}
      </option>
    );
  });

  return selected ? (
    <SelectedDropdown value={name} onChange={update}>
      {Title}
      {selectOptions}
    </SelectedDropdown>
  ) : (
    <OtherDropdown value={name} onChange={update}>
      {Title}
      {selectOptions}
    </OtherDropdown>
  );
}

TabList.propTypes = {
  options: PropTypes.array.isRequired,
  name: PropTypes.string.isRequired,
  selected: PropTypes.bool.isRequired,
  action: PropTypes.func
};

function TabItem({ name, children }) {
  return <span />;
}

TabItem.propTypes = {
  name: PropTypes.string.isRequired,
  children: PropTypes.oneOf([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ])
};

function TabContent({ children, ...props }) {
  return <ContentDiv>{children}</ContentDiv>;
}

TabContent.propTypes = {
  children: PropTypes.any.isRequired
};

class TabPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: ''
    };

    this.tabChanged = this.tabChanged.bind(this);
  }

  tabChanged = name => {
    this.setState({ value: name });
  };

  isChildSelected(value, index) {
    return value === this.state.value || (!this.state.value && index === 0);
  }

  render() {
    let content = null;
    const elements = React.Children.map(this.props.children, (child, i) => {
      let selected = false;
      if (child.type === Tab) {
        if (this.isChildSelected(child.props.name, i)) {
          content = <ContentDiv>{child.props.children}</ContentDiv>;
          selected = true;
        }
        return React.cloneElement(child, {
          selected,
          action: this.tabChanged,
          key: child.props.name
        });
      }
      if (child.type === TabList) {
        const options = React.Children.map(
          child.props.children,
          (secondChild, j) => {
            if (this.isChildSelected(secondChild.props.name, i)) {
              content = <ContentDiv>{secondChild.props.children}</ContentDiv>;
              selected = true;
            }
            return { optionText: secondChild.props.name };
          }
        );
        return React.cloneElement(child, {
          selected,
          action: this.tabChanged,
          options,
          key: child.props.name
        });
      }
      return null;
    });

    return (
      <div>
        {elements}
        {content}
      </div>
    );
  }
}

TabPanel.propTypes = {
  children: (props, propName, component) => {
    let badChild;
    const isTabItem = child => {
      const isTab = child.type.name === 'Tab' || child.type.name === 'TabList';
      if (!isTab) {
        badChild = child;
      }
      return isTab;
    };
    if (!props[propName].every(isTabItem)) {
      return new Error(
        `Tab Panel only accepts Tab or TabList as direct descendants. You gave a ${badChild
          .type.name || badChild.type}`
      );
    }
    return true;
  }
};

TabPanel.Tab = Tab;
TabPanel.TabList = TabList;
TabPanel.TabItem = TabItem;

export default TabPanel;
