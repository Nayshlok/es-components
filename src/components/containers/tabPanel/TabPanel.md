
Testing text

```
<TabPanel name="home">
  <TabPanel.Tab name="Home" value="test">
    <div>
      Hi there
    </div>
  </TabPanel.Tab>
  <TabPanel.Tab name="menu" value="testing">
    <div>
      Just another test
    </div>
  </TabPanel.Tab>
  <TabPanel.Tab name="Display" value="yetMoreTest">
    <div>
      Lipsom Epsum
    </div>
  </TabPanel.Tab>
</TabPanel>
```