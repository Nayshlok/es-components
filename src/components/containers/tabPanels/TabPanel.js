import React from 'react';
import PropTypes from 'prop-types';
import styled, { ThemeProvider } from 'styled-components';
import TabList from './TabList';
import Tab from './Tab';

const TabWrapper = styled('div')`
  display: flex;
  justify-content: center;

  @media (min-width: 899px) {
    justify-content: flex-start;
  }
`;

const TabFormatter = styled('div')`
  width: auto;
  display: flex;
  justify-content: center;
  flex-direction: column;

  @media (min-width: 899px) {
    flex-direction: row;
  }
`;

class TabPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
      currentContent: ''
    };
    this.tabChanged = this.tabChanged.bind(this);
  }

  componentWillMount() {
    if (this.state.value === '') {
      const child = this.props.children[0];
      const selectedChild = this.getFirstValueChild(child);
      this.tabChanged(selectedChild.props.name, selectedChild);
    }
  }

  getFirstValueChild(child) {
    if (Array.isArray(child.props.children)) {
      return child.props.children.filter(
        innerChild => innerChild.props.optionText
      )[0];
    }
    return child;
  }

  listIsSelected(list) {
    if (Array.isArray(list.props.children)) {
      return (
        list.props.children.filter(child => {
          if (child.props.optionText) {
            return (
              child.props.optionText.replace(/\s/g, '') === this.state.value
            );
          }
          return false;
        }).length > 0
      );
    }
    return false;
  }

  tabChanged(name, child) {
    this.setState({ value: name, currentContent: child });
  }

  render() {
    const { theme, children } = this.props;
    const elements = React.Children.map(children, (child, i) => {
      const isSelected =
        child.props.name === this.state.value ||
        this.listIsSelected(child) ||
        (i === 0 && this.state.value === '');
      return React.cloneElement(child, {
        key: child.props.name,
        selected: isSelected,
        action: this.tabChanged,
        selectedName: this.state.value
      });
    });

    return (
      <ThemeProvider theme={theme}>
        <div>
          <TabWrapper>
            <TabFormatter>{elements}</TabFormatter>
          </TabWrapper>
          <div>{this.state.currentContent}</div>
        </div>
      </ThemeProvider>
    );
  }
}

TabPanel.propTypes = {
  theme: PropTypes.object,
  children: PropTypes.node
};

TabPanel.defaultProps = {
  theme: {}
};

TabPanel.TabList = TabList;
TabPanel.Tab = Tab;

export default TabPanel;
