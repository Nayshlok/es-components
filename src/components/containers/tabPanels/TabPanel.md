Test Tab Panels

```
<div>
<TabPanel theme={{colors:{white: '#000'}}}>
  <TabPanel.Tab name="test1">
    <p>
      HELLO WORLD!!!!
    </p>
  </TabPanel.Tab>
  <TabPanel.Tab name="test2">
    <div>
      <p>
        Nesting is fun
      </p>
    </div>
  </TabPanel.Tab>
  <TabPanel.TabList name="testList">
    <span optionText="test3">
       No elements
    </span>
    <span optionText="test4">
      <p>
        Another
      </p>
    </span>
  </TabPanel.TabList>
</TabPanel>
</div>
```