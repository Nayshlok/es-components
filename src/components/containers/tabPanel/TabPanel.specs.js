/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import TabPanel from './TabPanel';

const buildTabPanel = props => (
  <TabPanel>
    <TabPanel.Tab name="tab1">first</TabPanel.Tab>
    <TabPanel.Tab name="tab2" className="testing">
      second
    </TabPanel.Tab>
    <TabPanel.Tab name="tab3">third</TabPanel.Tab>
    <TabPanel.TabList name="multiTab">
      <TabPanel.TabItem name="option1" value="TheFirst">
        <div>multi first</div>
      </TabPanel.TabItem>
      <TabPanel.TabItem name="option2" value="TheSecond">
        <div>multi second</div>
      </TabPanel.TabItem>
      <TabPanel.TabItem name="option3" value="TheThird">
        <div>multi third</div>
      </TabPanel.TabItem>
    </TabPanel.TabList>
    <div />
  </TabPanel>
);

describe('Tab panel', () => {
  it('renders as expected', () => {
    const tree = renderer.create(buildTabPanel()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('clicking a tab should set the panels value to its name', () => {
    const panelInstance = mount(buildTabPanel());
    const secondTab = panelInstance.find('.testing');
    expect(panelInstance.state('value')).toBe('');
    secondTab.simulate('click');
    expect(panelInstance.state('value')).toBe('tab2');
  });
});
