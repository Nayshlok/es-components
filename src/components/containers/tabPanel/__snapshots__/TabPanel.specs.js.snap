// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab panel renders as expected 1`] = `
<div>
  <button
    className="sc-bdVaJa eYyOtq"
  >
    tab1
  </button>
  <button
    className="testing sc-bwzfXH dnBPMW"
    onClick={[Function]}
  >
    tab2
  </button>
  <button
    className="sc-bwzfXH dnBPMW"
    onClick={[Function]}
  >
    tab3
  </button>
  <select
    className="sc-bxivhb eOMafC"
    onChange={[Function]}
    value="multiTab"
  >
    <option
      disabled={true}
    >
      multiTab
    </option>
    <option
      value="option1"
    >
      option1
    </option>
    <option
      value="option2"
    >
      option2
    </option>
    <option
      value="option3"
    >
      option3
    </option>
  </select>
  <div
    className="sc-ifAKCX hjIvEj"
  >
    first
  </div>
</div>
`;
